File: internal/parser/analyzer.go
package parser

import (
	"regexp"
	"strconv"
	"strings"

	"longevity-ranker/internal/models"
	"longevity-ranker/internal/rules"
)

var (
	reMg      = regexp.MustCompile(`(?i)(\d+)\s*mg`)
	reCount   = regexp.MustCompile(`(?i)(\d+)\s*(?:capsules|caps|servings|tabs|tablets|ct)`)
	reGrams   = regexp.MustCompile(`(?i)(\d+)\s*(?:grams?|gms?|g)\b`)
	reKg      = regexp.MustCompile(`(?i)(\d+(?:\.\d+)?)\s*kg\b`)
	rePack    = regexp.MustCompile(`(?i)(\d+)\s*(?:Pack|Bottles?)`)
	reServing = regexp.MustCompile(`(?i)(\d+)\s*(?:capsules|caps).*?per\s*serving`)

	// reLabelGrams and reLabelKg scan only variant.Title and product.Title (label text)
	// for Gross Grams extraction. Identical patterns to reGrams/reKg but kept separate
	// for clarity of intent.
	reLabelGrams = regexp.MustCompile(`(?i)(\d+)\s*(?:grams?|gms?|g)\b`)
	reLabelKg    = regexp.MustCompile(`(?i)(\d+(?:\.\d+)?)\s*kg\b`)
)

// dirtyKeywords flags products whose regex-extracted mass is likely unreliable.
var dirtyKeywords = []string{
	"flavor", "island cooler", "coastal explosion", "watermelon", "berry", "punch",
	"orange", "lemon", "mango", "grape", "apple", "blend", "complex", "with", "+",
	"gumm", "chew", "bundle", "blue raspberry", "fruit punch", "sour watermelon",
	"pineapple mango", "mandarin orange", "shaq's berry blast", "frozen lemonade",
}

// Analyzer holds the configuration needed by the analysis and audit pipelines.
// There is no global mutable state — all dependencies are injected here.
type Analyzer struct {
	Rules       rules.Registry
	Supplements []string
}

// matchesSupplement reports whether the product's identity string contains at
// least one of the configured supplement keywords.
func (a *Analyzer) matchesSupplement(identity string) bool {
	return containsAny(identity, a.Supplements)
}

// vendorConfig returns the VendorConfig for the given vendor name, plus the
// product-level spec and whether an override exists for the given handle.
func (a *Analyzer) vendorConfig(vendorName, handle string) (cfg rules.VendorConfig, spec rules.ProductSpec, hasOverride bool) {
	if a.Rules == nil {
		return
	}
	cfg, exists := a.Rules[vendorName]
	if !exists {
		return
	}
	spec, hasOverride = cfg.Overrides[handle]
	return
}

// AnalyzeProduct evaluates every available variant of a product and returns an
// Analysis entry for each valid one. It implements a Hybrid Catalog/Regex Engine:
//
//   - If the product handle has an override with ForceActiveGrams > 0, the regex
//     mass-extraction pipeline is bypassed entirely.
//   - The pack multiplier regex (rePack) always runs regardless of overrides.
//   - When GlobalSubscriptionDiscount is configured, a synthetic "Subscribe & Save"
//     entry is emitted for each variant.
//
// Returns nil when the product has no variants, does not match any allowed
// supplement keyword, or yields no valid analyses.
func (a *Analyzer) AnalyzeProduct(vendorName string, p models.Product) []models.Analysis {
	if len(p.Variants) == 0 {
		return nil
	}

	identity := strings.ToLower(p.Title + " " + p.Context + " " + p.Handle)
	if !a.matchesSupplement(identity) {
		return nil
	}

	cfg, spec, hasOverride := a.vendorConfig(vendorName, p.Handle)

	var results []models.Analysis

	for _, v := range p.Variants {
		if !v.Available {
			continue
		}

		// Variant-level blocklist
		if len(cfg.VariantBlocklist) > 0 && containsAny(strings.ToLower(v.Title), cfg.VariantBlocklist) {
			continue
		}

		price, err := strconv.ParseFloat(v.Price, 64)
		if err != nil || price <= 0 {
			continue
		}

		// --- Search strings at different specificity levels ---
		variantSearch := v.Title
		cleanSearch := p.Title + " " + v.Title
		broadSearch := p.Title + " " + p.Context + " " + v.Title + " " + strings.ReplaceAll(p.Handle, "-", " ") + " " + p.BodyHTML

		// =================================================================
		// ACTIVE GRAMS EXTRACTION — Hybrid Engine
		// =================================================================
		capsuleMass, powderMass, usedOverride := a.extractMass(spec, hasOverride, v.Title, cleanSearch, broadSearch, variantSearch)

		baseMass := capsuleMass + powderMass

		// =================================================================
		// PACK MULTIPLIER — Always runs regardless of override source
		// =================================================================
		packMultiplier := 1.0
		if m, ok := extractFloatFrom(rePack, variantSearch, broadSearch); ok {
			packMultiplier = m
		}

		activeGrams := baseMass * packMultiplier
		if activeGrams <= 0 {
			continue
		}

		// =================================================================
		// GROSS GRAMS EXTRACTION — Label Weight
		// =================================================================
		isCapsuleProduct := capsuleMass > 0 && powderMass == 0
		grossGrams := a.extractGrossGrams(spec, hasOverride, v.Title, p.Title, isCapsuleProduct, packMultiplier)

		// =================================================================
		// PURE POWDER FALLBACK
		// =================================================================
		if !usedOverride && grossGrams > 0 && !isCapsuleProduct {
			triageTarget := strings.ToLower(p.Title + " " + v.Title + " " + p.Handle)
			if !containsAny(triageTarget, dirtyKeywords) {
				activeGrams = grossGrams
			}
		}

		// =================================================================
		// TYPE DETERMINATION — Hybrid Engine
		// =================================================================
		typeSearch := strings.ToLower(p.Title + " " + v.Title + " " + p.Handle + " " + p.Context)
		productType := classifyType(typeSearch, spec, hasOverride, usedOverride, packMultiplier, capsuleMass, powderMass)

		// --- Bioavailability multiplier ---
		multiplier, multiplierLabel := bioavailabilityMultiplier(typeSearch, productType)

		// --- Display name ---
		displayName := buildDisplayName(p.Title, v.Title, vendorName)

		// =================================================================
		// TRIAGE ENGINE — Dirty Data Detection
		// =================================================================
		needsReview, reviewReason := a.triageDirtyData(usedOverride, displayName, p.Handle, p.Title)

		// Pure powder gross fallback
		if productType == "Powder" && grossGrams == 0 && !needsReview {
			grossGrams = activeGrams
		}

		// --- One-time purchase entry ---
		results = append(results, buildAnalysis(
			vendorName, displayName, p.Handle, p.ImageURL, productType,
			price, activeGrams, grossGrams, multiplier, multiplierLabel,
			false, needsReview, reviewReason,
		))

		// --- Synthetic subscription entry ---
		if cfg.GlobalSubscriptionDiscount > 0 {
			subPrice := price * (1 - cfg.GlobalSubscriptionDiscount)
			results = append(results, buildAnalysis(
				vendorName, displayName+" (Subscribe & Save)", p.Handle, p.ImageURL, productType,
				subPrice, activeGrams, grossGrams, multiplier, multiplierLabel,
				true, needsReview, reviewReason,
			))
		}
	}

	if len(results) == 0 {
		return nil
	}
	return results
}

// extractMass implements the hybrid catalog/regex mass-extraction pipeline.
// Returns capsuleMass, powderMass, and whether an override was used.
func (a *Analyzer) extractMass(spec rules.ProductSpec, hasOverride bool, variantTitle, cleanSearch, broadSearch, variantSearch string) (capsuleMass, powderMass float64, usedOverride bool) {
	// VARIANT CATALOG PATH
	if hasOverride && spec.VariantOverrides != nil && spec.VariantOverrides[variantTitle] > 0 {
		return 0, spec.VariantOverrides[variantTitle], true
	}

	// PRODUCT CATALOG PATH
	if hasOverride && spec.ForceActiveGrams > 0 {
		return 0, spec.ForceActiveGrams, true
	}

	// REGEX PATH

	// Step 1: Explicit grams or kg in clean title+variant
	if g, ok := extractFloat(reGrams, cleanSearch); ok {
		return 0, g, false
	}
	if kg, ok := extractFloat(reKg, cleanSearch); ok {
		return 0, kg * 1000.0, false
	}

	// Step 2: mg × count (capsules/tablets)
	mg, mgOk := extractFloat(reMg, broadSearch)
	count, countOk := extractFloatFrom(reCount, variantSearch, cleanSearch, broadSearch)
	if mgOk && countOk {
		servingSize := 1.0
		if s, ok := extractFloat(reServing, broadSearch); ok {
			servingSize = s
		}
		capsuleMass = (mg / servingSize * count) / 1000.0
		return capsuleMass, 0, false
	}

	// Step 3: Fallback — grams in broad search
	if g, ok := extractFloat(reGrams, broadSearch); ok {
		return 0, g, false
	}

	return 0, 0, false
}

// extractGrossGrams extracts the physical label weight from variant/product titles.
func (a *Analyzer) extractGrossGrams(spec rules.ProductSpec, hasOverride bool, variantTitle, productTitle string, isCapsule bool, packMult float64) float64 {
	// Variant-level gross override
	if hasOverride && spec.VariantGrossOverrides != nil && spec.VariantGrossOverrides[variantTitle] > 0 {
		return spec.VariantGrossOverrides[variantTitle]
	}

	// Capsules don't have a meaningful gross weight
	if isCapsule {
		return 0
	}

	labelSearch := productTitle + " " + variantTitle
	if g, ok := extractFloat(reLabelGrams, labelSearch); ok {
		return g * packMult
	}
	if kg, ok := extractFloat(reLabelKg, labelSearch); ok {
		return kg * 1000.0 * packMult
	}
	return 0
}

// classifyType determines the product type string.
func classifyType(typeSearch string, spec rules.ProductSpec, hasOverride, usedOverride bool, packMult, capsuleMass, powderMass float64) string {
	if hasOverride && spec.ForceType != "" {
		return spec.ForceType
	}
	if packMult > 1 {
		return "Multi-Pack"
	}
	if !usedOverride && capsuleMass > 0 && powderMass > 0 {
		return "Hybrid Bundle"
	}
	if !usedOverride && powderMass > 0 && capsuleMass == 0 {
		return "Powder"
	}
	if strings.Contains(typeSearch, "gel") && !strings.Contains(typeSearch, "softgel") {
		return "Gel"
	}
	if strings.Contains(typeSearch, "tab") {
		return "Tablets"
	}
	if strings.Contains(typeSearch, "powder") {
		return "Powder"
	}
	return "Capsules"
}

// bioavailabilityMultiplier returns the multiplier and label for delivery bonuses.
func bioavailabilityMultiplier(typeSearch, productType string) (float64, string) {
	switch {
	case strings.Contains(typeSearch, "liposomal") || strings.Contains(typeSearch, "lipo"):
		return 1.5, "Lipo Bonus"
	case strings.Contains(typeSearch, "sublingual"):
		return 1.1, "Sublingual"
	case productType == "Gel":
		return 1.1, "Gel Bonus"
	case productType == "Tablets":
		return 1.1, "Tablet Bonus"
	default:
		return 1.0, ""
	}
}

// buildDisplayName constructs the user-facing product name, stripping the
// redundant vendor name prefix and appending the variant title when meaningful.
func buildDisplayName(productTitle, variantTitle, vendorName string) string {
	name := productTitle
	if variantTitle != "" && !strings.EqualFold(variantTitle, "Default Title") {
		name += " (" + variantTitle + ")"
	}

	// Strip redundant vendor name prefix (case-insensitive)
	if len(vendorName) > 0 && len(name) >= len(vendorName) &&
		strings.EqualFold(name[:len(vendorName)], vendorName) {
		if trimmed := strings.TrimSpace(name[len(vendorName):]); len(trimmed) > 0 {
			name = trimmed
		}
	}
	return name
}

// triageDirtyData checks whether regex-extracted mass is likely unreliable.
func (a *Analyzer) triageDirtyData(usedOverride bool, displayName, handle, title string) (bool, string) {
	if usedOverride {
		return false, ""
	}

	triageTarget := strings.ToLower(displayName + " " + handle + " " + title)
	for _, kw := range dirtyKeywords {
		if !strings.Contains(triageTarget, strings.ToLower(kw)) {
			continue
		}
		// "unflavored" contains substring "flavor" but is not a dirty signal
		if kw == "flavor" && strings.Contains(triageTarget, "unflavored") {
			if strings.Contains(triageTarget, "serv") {
				return true, "Detected 'unflavored' but uses 'servings' (needs manual math check)"
			}
			continue
		}
		return true, "Detected dirty keyword: " + kw
	}
	return false, ""
}

// buildAnalysis constructs a single Analysis entry with computed cost metrics.
func buildAnalysis(
	vendor, name, handle, imageURL, productType string,
	price, activeGrams, grossGrams, multiplier float64, multiplierLabel string,
	isSubscription, needsReview bool, reviewReason string,
) models.Analysis {
	costPerGram := price / activeGrams
	return models.Analysis{
		Vendor:          vendor,
		Name:            name,
		Handle:          handle,
		Price:           price,
		ActiveGrams:     activeGrams,
		GrossGrams:      grossGrams,
		CostPerGram:     costPerGram,
		EffectiveCost:   costPerGram / multiplier,
		Multiplier:      multiplier,
		MultiplierLabel: multiplierLabel,
		Type:            productType,
		ImageURL:        imageURL,
		IsSubscription:  isSubscription,
		NeedsReview:     needsReview,
		ReviewReason:    reviewReason,
	}
}

File: web/lib/analyzer.ts
/**
 * TypeScript port of internal/parser/analyzer.go
 *
 * Replicates the Go math engine: regex extraction of mg, count, grams, kg,
 * pack multiplier, serving size, type classification, bioavailability multiplier,
 * and effective cost calculation.
 *
 * Every regex and priority order matches the Go implementation exactly.
 */

import type { Product, Variant, Analysis, VendorRules } from "./types";

// --- Regexes (mirrors Go `var` block) ---

const reMg = /(\d+)\s*mg/i;
const reCount = /(\d+)\s*(?:capsules|caps|servings|tabs|tablets|ct)/i;
const reGrams = /(\d+)\s*(?:grams?|g)\b/i;
const reKg = /(\d+(?:\.\d+)?)\s*kg\b/i;
const rePack = /(\d+)\s*Pack/i;
const reServing = /(\d+)\s*(?:capsules|caps).*?per\s*serving/i;

// --- Allowed supplements (mirrors Go AllowedSupplements) ---

const DEFAULT_SUPPLEMENTS = [
  "nmn",
  "nad",
  "tmg",
  "trimethylglycine",
  "resveratrol",
  "creatine",
];

/**
 * extractCount replicates the Go helper of the same name.
 * It tries progressively broader search strings and returns the first match.
 */
function extractCount(
  variantSearch: string,
  cleanSearch: string,
  broadSearch: string
): RegExpMatchArray | null {
  // Priority 1: variant title alone
  let m = variantSearch.match(reCount);
  if (m) return m;

  // Priority 2: product title + variant title
  m = cleanSearch.match(reCount);
  if (m) return m;

  // Priority 3: everything
  m = broadSearch.match(reCount);
  if (m) return m;

  return null;
}

/**
 * applyRules replicates rules.ApplyRules from Go.
 * Returns false if the product should be rejected (blocklisted).
 * Mutates `product.context` in place when overrides match.
 */
export function applyRules(
  vendorName: string,
  product: Product,
  rules: VendorRules
): boolean {
  const config = rules[vendorName];
  if (!config) return true;

  // 1. Check blocklist
  const identity = (
    product.title +
    " " +
    product.handle +
    " " +
    product.context
  ).toLowerCase();

  for (const blocked of config.blocklist ?? []) {
    if (identity.includes(blocked.toLowerCase())) {
      return false;
    }
  }

  // 2. Apply overrides
  const spec = config.overrides?.[product.handle];
  if (spec) {
    let extraContext = "";
    if (spec.forceMg > 0) {
      extraContext += ` ${spec.forceMg}mg`;
    }
    if (spec.forceCount > 0) {
      if (spec.forceType === "Powder") {
        extraContext += ` ${spec.forceCount}g`;
      } else if (spec.forceType === "Tablets") {
        extraContext += ` ${spec.forceCount} tablets`;
      } else {
        extraContext += ` ${spec.forceCount} capsules`;
      }
    }
    product.context += extraContext;
  }

  return true;
}

/**
 * analyzeProduct is a line-for-line port of Go's AnalyzeProduct.
 *
 * Given a vendor name and product, it iterates over available variants,
 * extracts dosage information from progressively broader search strings,
 * computes cost-per-gram and effective cost (with bioavailability multiplier),
 * and returns the best (lowest $/g) analysis — or null if nothing matched.
 */
export function analyzeProduct(
  vendorName: string,
  product: Product,
  allowedSupplements: string[] = DEFAULT_SUPPLEMENTS
): Analysis | null {
  if (!product.variants || product.variants.length === 0) {
    return null;
  }

  // --- Supplement keyword filter ---
  const identityString = (
    product.title +
    " " +
    product.context +
    " " +
    product.handle
  ).toLowerCase();

  let matched = false;
  for (const supp of allowedSupplements) {
    if (identityString.includes(supp)) {
      matched = true;
      break;
    }
  }
  if (!matched) return null;

  let bestAnalysis: Analysis | null = null;
  let minCostPerGram = Number.MAX_VALUE;

  for (const v of product.variants) {
    if (!v.available) continue;

    const price = parseFloat(v.price);
    if (!price || price <= 0) continue;

    // --- Build search strings at different specificity levels ---
    const variantSearch = v.title;
    const cleanSearch = product.title + " " + v.title;
    const broadSearch =
      product.title +
      " " +
      product.context +
      " " +
      v.title +
      " " +
      product.handle.replace(/-/g, " ") +
      " " +
      product.body_html;

    let capsuleMass = 0;
    let powderMass = 0;
    let packMultiplier = 1;

    // --- Step 1: Check for explicit grams or kg in clean title+variant ---
    const gramMatch = cleanSearch.match(reGrams);
    const kgMatch = cleanSearch.match(reKg);

    if (gramMatch) {
      powderMass = parseFloat(gramMatch[1]);
    } else if (kgMatch) {
      powderMass = parseFloat(kgMatch[1]) * 1000;
    } else {
      // --- Step 2: Extract mg and capsule count ---
      const mgMatch = broadSearch.match(reMg);
      const countMatch = extractCount(variantSearch, cleanSearch, broadSearch);

      if (mgMatch && countMatch) {
        const mg = parseFloat(mgMatch[1]);
        const count = parseFloat(countMatch[1]);

        const servingMatch = broadSearch.match(reServing);
        let servingSize = 1;
        if (servingMatch) {
          const s = parseFloat(servingMatch[1]);
          if (s > 0) servingSize = s;
        }

        capsuleMass = (mg / servingSize * count) / 1000;
      }
    }

    // --- Step 3: Fallback — check broad search for grams ---
    if (powderMass === 0 && capsuleMass === 0) {
      const gramMatchBody = broadSearch.match(reGrams);
      if (gramMatchBody) {
        powderMass = parseFloat(gramMatchBody[1]);
      }
    }

    // --- Step 4: Pack multiplier ---
    let packMatch = variantSearch.match(rePack);
    if (!packMatch) {
      packMatch = broadSearch.match(rePack);
    }
    if (packMatch) {
      packMultiplier = parseFloat(packMatch[1]);
    }

    const totalGrams = (capsuleMass + powderMass) * packMultiplier;
    if (totalGrams <= 0) continue;

    const costPerGram = price / totalGrams;

    // --- Type classification (never uses BodyHTML to avoid HTML tag leakage) ---
    const typeSearch = (
      product.title +
      " " +
      v.title +
      " " +
      product.handle +
      " " +
      product.context
    ).toLowerCase();

    let productType = "Single";

    if (packMultiplier > 1) {
      productType = "Multi-Pack";
    } else if (capsuleMass > 0 && powderMass > 0) {
      productType = "Hybrid Bundle";
    } else if (powderMass > 0) {
      productType = "Powder";
    } else if (
      typeSearch.includes("gel") &&
      !typeSearch.includes("softgel")
    ) {
      productType = "Gel";
    } else if (typeSearch.includes("tab")) {
      productType = "Tablets";
    } else {
      productType = "Capsules";
    }

    // --- Bioavailability multiplier ---
    let multiplier = 1.0;

    if (
      typeSearch.includes("liposomal") ||
      typeSearch.includes("lipo")
    ) {
      multiplier = 1.5;
    } else if (
      typeSearch.includes("sublingual") ||
      productType === "Gel" ||
      productType === "Tablets"
    ) {
      multiplier = 1.1;
    }

    const effectiveCost = costPerGram / multiplier;

    if (costPerGram < minCostPerGram) {
      minCostPerGram = costPerGram;

      let displayName = product.title;
      if (
        v.title &&
        v.title.toLowerCase() !== "default title"
      ) {
        displayName = displayName + " (" + v.title + ")";
      }

      bestAnalysis = {
        vendor: vendorName,
        name: displayName,
        price,
        totalGrams,
        costPerGram,
        effectiveCost,
        type: productType,
        imageURL: product.image_url,
        handle: product.handle,
      };
    }
  }

  return bestAnalysis;
}

/**
 * Analyze all products for a single vendor, applying rules first.
 * Returns an array of Analysis results (one per qualifying product).
 */
export function analyzeVendor(
  vendorName: string,
  products: Product[],
  rules: VendorRules,
  allowedSupplements?: string[]
): Analysis[] {
  const results: Analysis[] = [];

  for (const product of products) {
    // Deep-clone so context mutation from applyRules doesn't pollute the source
    const p: Product = JSON.parse(JSON.stringify(product));

    if (!applyRules(vendorName, p, rules)) continue;

    const analysis = analyzeProduct(vendorName, p, allowedSupplements);
    if (analysis) {
      results.push(analysis);
    }
  }

  return results;
}

/**
 * Analyze all vendors at once. Returns a flat array sorted by effectiveCost ascending.
 */
export function analyzeAll(
  vendorProducts: { vendorName: string; products: Product[] }[],
  rules: VendorRules,
  allowedSupplements?: string[]
): Analysis[] {
  const all: Analysis[] = [];

  for (const { vendorName, products } of vendorProducts) {
    all.push(...analyzeVendor(vendorName, products, rules, allowedSupplements));
  }

  all.sort((a, b) => a.effectiveCost - b.effectiveCost);
  return all;
}